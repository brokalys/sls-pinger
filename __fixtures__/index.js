export * from './pinger';
export * from './property';
