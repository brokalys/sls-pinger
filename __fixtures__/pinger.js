export function createPingerFixture(customData = {}) {
  return {
    id: 1,
    categories: ['apartment'],
    types: ['sell'],
    price_min: 100000,
    price_max: 200000,
    price_type: 'total',
    rooms_min: null,
    rooms_max: null,
    area_m2_min: null,
    area_m2_max: null,
    floor_min: null,
    floor_max: null,
    location:
      '56.992294 24.136619, 56.976394 23.995790, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176, 56.992294 24.136619, 56.992294 24.136619',
    frequency: 'immediate',
    is_premium: true,
    ...customData,
  };
}
