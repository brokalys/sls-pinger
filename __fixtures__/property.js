function createPropertyFixture(customData = {}) {
  return {
    category: 'apartment',
    type: 'sell',
    lat: 56.9965,
    lng: 24.245176,
    price: 100000,
    url: 'https://brokalys.com/',
    area: 45,
    rooms: 2,
    calc_price_per_sqm: 122.32,
    ...customData,
  };
}

function createPropertyQueueItemFixture(customData = {}) {
  return {
    id: 1,
    pinger_id: 1,
    locked: 0,
    ...customData,
    data: JSON.stringify(createPropertyFixture(customData.data)),
  };
}

function createPropertyStatFixture(customData = {}) {
  return {
    id: 1,
    pinger_id: 1,
    data: ['120.00', '200.20'],
    created_at: '2021-01-01T10:20:22Z',
    ...customData,
  };
}

module.exports = {
  createPropertyFixture,
  createPropertyQueueItemFixture,
  createPropertyStatFixture,
};
