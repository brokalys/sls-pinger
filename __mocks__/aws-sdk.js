export const sns = {
  publish: jest.fn().mockReturnValue({ promise: jest.fn() }),
};

const aws = {
  SNS: jest.fn().mockReturnValue(sns),
};

export default aws;
