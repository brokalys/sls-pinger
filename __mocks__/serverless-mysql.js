const query = jest.fn().mockReturnValue([]);

const mysql = jest.fn().mockReturnValue({
  query,
});

module.exports = {
  query,
  default: mysql,
};
