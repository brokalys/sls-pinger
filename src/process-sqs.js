const numeral = require('numeral');
const inside = require('point-in-polygon');
const db = require('./shared/db');
const sns = require('./shared/sns');
const createUnsubscribeLink = require('./shared/unsubscribe-link');
const utils = require('./shared/utils');

function parseLocation(location) {
  return location
    .split(', ')
    .map((row) => row.split(' ').map((row) => parseFloat(row)))
    .slice(0, -1);
}

function nl2br(str, is_xhtml) {
  const breakTag =
    is_xhtml || typeof is_xhtml === 'undefined' ? '<br />' : '<br>';
  return (str + '').replace(
    /([^>\r\n]?)(\r\n|\n\r|\r|\n)/g,
    '$1' + breakTag + '$2',
  );
}

function parseImages(img) {
  if (!img) {
    return [];
  }

  if (img instanceof Array) {
    return img;
  }

  return JSON.parse(img);
}

exports.run = async (event, context) => {
  const properties = event.Records.map((row) => JSON.parse(row.body)).filter(
    (property) =>
      property.lat &&
      property.lng &&
      property.price > 0 &&
      ['sell', 'rent'].includes(property.type),
  );

  if (properties.length === 0) {
    return;
  }

  const results = await db.getAvailablePingers();

  const invocations = properties.map((property) => {
    const pingers = results
      .filter((pinger) => pinger.categories.includes(property.category))
      .filter((pinger) => pinger.types.includes(property.type))
      .filter(
        (pinger) =>
          pinger.price_min === null ||
          (pinger.price_type === 'sqm'
            ? property.calc_price_per_sqm >= pinger.price_min
            : property.price >= pinger.price_min),
      )
      .filter(
        (pinger) =>
          pinger.price_max === null ||
          (pinger.price_type === 'sqm'
            ? property.calc_price_per_sqm <= pinger.price_max
            : property.price <= pinger.price_max),
      )
      .filter(
        (pinger) =>
          pinger.rooms_min === null || property.rooms >= pinger.rooms_min,
      )
      .filter(
        (pinger) =>
          pinger.rooms_max === null || property.rooms <= pinger.rooms_max,
      )
      .filter(
        (pinger) =>
          pinger.area_m2_min === null || property.area >= pinger.area_m2_min,
      )
      .filter(
        (pinger) =>
          pinger.area_m2_max === null || property.area <= pinger.area_m2_max,
      )
      .filter(
        (pinger) =>
          pinger.floor_min === null || property.floor >= pinger.floor_min,
      )
      .filter(
        (pinger) =>
          pinger.floor_max === null || property.floor <= pinger.floor_max,
      )
      .filter((pinger) =>
        inside([property.lat, property.lng], parseLocation(pinger.location)),
      );

    return pingers.map((pinger) => ({ ...pinger, property }));
  });

  // Flatten the data
  const availableInvocations = [].concat.apply([], invocations);

  await Promise.all(
    availableInvocations
      .map((pinger) => {
        const result = { ...pinger.property };
        result.images = parseImages(result.images);
        result.content = nl2br(
          (result.content || '').replace(/(<([^>]+)>)/gi, ''),
        );

        result.unsubscribe_url = createUnsubscribeLink(pinger);
        result.url = `https://view.brokalys.com/?link=${encodeURIComponent(
          result.url,
        )}`;
        result.price = numeral(result.price).format('0,0 €');
        result.is_premium = !!pinger.is_premium;

        return {
          to: pinger.email,
          pinger_id: pinger.id,
          template_id: 'email',
          template_variables: result,
          frequency: pinger.frequency,
        };
      })
      .map((data) => performAction(data, context)),
  );
};

function performAction(data, context) {
  if (data.frequency === 'immediate') {
    return publishSns(data, context);
  }

  return queuePinger(data);
}

function publishSns(data, context) {
  return sns
    .publish({
      Message: 'email',
      MessageAttributes: {
        to: {
          DataType: 'String',
          StringValue: data.to,
        },
        subject: {
          DataType: 'String',
          StringValue: 'Jauns PINGER sludinājums',
        },
        pinger_id: {
          DataType: 'Number',
          StringValue: String(data.pinger_id),
        },
        template_id: {
          DataType: 'String',
          StringValue: data.template_id,
        },
        template_variables: {
          DataType: 'String',
          StringValue: JSON.stringify(data.template_variables),
        },
      },
      MessageStructure: 'string',
      TargetArn: utils.constructArn(context, process.env.EMAIL_SNS_TOPIC_NAME),
    })
    .promise();
}

function queuePinger(data) {
  return db.queuePingerForSummaryEmail(data.pinger_id, data.template_variables);
}
