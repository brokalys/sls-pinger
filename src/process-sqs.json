{
  "Records": [
    {
      "body": "{\"url\":\"https://inch.lv/details/apartments/11408\",\"type\":\"sell\",\"category\":\"apartment\",\"source\":\"inch.lv\",\"content\":\"Īpašnieks pārdod stilīgu un mājīgu dzīvokli pirmā augstā stāvā, Rīgas centrā. Dzīvoklis ir ideālā stāvoklī, grīdas segums - parkets, augstie griesti - 3, 3 m, uzstādīta signalizācija. Viesistaba apvienota ar virtuvi (38, 4 m²), divas izolētas guļamistabas (12, 9 м², 16, 2 м²). Divas vannas istabas ir gan vanna gan duša, siltās grīdas. Dzīvoklis ir mēbelēts un aprīkots ar sadzīves tehniku (televizors, elektr. plīts, trauku mazgājamā mašīna, cepeškrāsnis, ledusskapis, veļas mašīna). Visas ērtības, apkure Rīgas Siltums. Zemes domājama dāļa īpašumā. Viena autostāvvieta slēgtā pagalmā ar automātiskiem vārtiem. \\n\\nСобственник продает стильную и уютную квартиру в центре Риги, высокий первый этаж. Квартира в идеальном состоянии, напольное покрытие - паркет, высокие потолки – 3, 3 м, установлена сигнализация. Гостинная совмещена с кухней (38, 4 м²), две изолированные спальни (12, 9 м², 16, 2 м²). Два санузла: в первом установленна ванна, во втором душевая кабина, полы в санузлах с подогревом. Квартира полностью мебелирована и оборудована бытовой техникой (телевизор, электр. плита, посудомоечная машина, духовка, холодильник, стиральная машина). Все удобства, отопление Rīgas Siltums. Земля в собственности. Одно парковочное место в закрытом дворе, с автоматическими воротами.\",\"location_district\":\"Rīga\",\"location_parish\":\"Centrs\",\"location_address\":\"Cēsu iela, Видземское предместье, Рига, Латвия, 43\",\"area\":71,\"price_per_sqm\":1661.9718,\"price\":118000,\"rooms\":3,\"lat\":24.1313266301025,\"lng\":56.9623341618612,\"floor\":1,\"max_floors\":4,\"contact_phone\":\"+37129628479\",\"foreign_id\":\"11408\",\"images\":\"[\\\"https://i.inch.lv/images/apartment/11408/full_image0.jpg?t=1592769540\\\",\\\"https://i.inch.lv/images/apartment/11408/full_image1.jpg?t=1592769540\\\",\\\"https://i.inch.lv/images/apartment/11408/full_image2.jpg?t=1592769540\\\",\\\"https://i.inch.lv/images/apartment/11408/full_image3.jpg?t=1592769540\\\",\\\"https://i.inch.lv/images/apartment/11408/full_image4.jpg?t=1592769540\\\"]\",\"building_project\":\"pre-war\",\"published_at\":\"2020-06-21T19:58:52.308Z\",\"additional_data\":\"{\\\"email\\\":\\\"cranio247@inbox.lv\\\",\\\"name\\\":\\\"Irina\\\",\\\"raw\\\":{\\\"id\\\":11408,\\\"city\\\":\\\"Rīga\\\",\\\"district\\\":\\\"Centrs\\\",\\\"address\\\":\\\"Cēsu iela, Видземское предместье, Рига, Латвия, 43\\\",\\\"longitude\\\":24.1313266301025,\\\"latitude\\\":56.9623341618612,\\\"dealType\\\":\\\"sale\\\",\\\"area\\\":71,\\\"houseSeries\\\":\\\"pre-war\\\",\\\"roomCount\\\":3,\\\"floorNumber\\\":1,\\\"floorTotal\\\":4,\\\"hasElevator\\\":false,\\\"hasParking\\\":true,\\\"hasBalcony\\\":false,\\\"primaryPhone\\\":\\\"+37129628479\\\",\\\"secondaryPhone\\\":\\\"27700588\\\",\\\"email\\\":\\\"cranio247@inbox.lv\\\",\\\"userCreatedAt\\\":\\\"2020-06-21T19:58:52.308Z\\\",\\\"userUpdatedAt\\\":\\\"2020-06-21T19:58:52.308Z\\\",\\\"pricePerArea\\\":1661.9718,\\\"specialCondition\\\":\\\"flexible\\\",\\\"searchShowCount\\\":2,\\\"type\\\":\\\"apartment\\\",\\\"price\\\":118000,\\\"processingImages\\\":false,\\\"fullName\\\":\\\"Irina\\\"}}\",\"rent_type\":\"unknown\",\"image_count\":5}"
    }
  ]
}
