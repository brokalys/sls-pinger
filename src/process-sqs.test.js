import { query } from 'serverless-mysql';
import { sns } from 'aws-sdk';
import { run } from './process-sqs';
import { createPingerFixture, createPropertyFixture } from '__fixtures__';

jest.mock('aws-sdk');
jest.mock('serverless-mysql');

function createMockRecords(records) {
  return records.map((data) => ({
    body: JSON.stringify(data),
  }));
}

describe('process-sqs', () => {
  afterEach(jest.clearAllMocks);

  test('publishes a new SNS message if PINGER falls within defined bounds', async () => {
    const event = {
      Records: createMockRecords([createPropertyFixture()]),
    };
    query.mockReturnValue([
      createPingerFixture(),
      createPingerFixture({
        rooms_min: 4,
      }),
    ]);

    await run(event);

    expect(sns.publish).toBeCalledTimes(1);
    expect(query).toBeCalledTimes(1);
  });

  test.each(['daily', 'weekly', 'monthly'])(
    'writes a new queue entry if pinger invocation is %j',
    async (type) => {
      const event = {
        Records: createMockRecords([createPropertyFixture()]),
      };
      query.mockReturnValue([
        createPingerFixture({ type }),
        createPingerFixture({
          rooms_min: 4,
          type,
        }),
      ]);

      await run(event);

      expect(sns.publish).not.toBeCalled();
      expect(query).toBeCalledWith(
        expect.objectContaining({
          sql: expect.stringContaining('INSERT INTO pinger_queue'),
        }),
      );
    },
  );

  test('does not publish an SNS message if PINGER falls outside of defined bounds', async () => {
    const event = {
      Records: createMockRecords([createPropertyFixture({ lat: 59.9965 })]),
    };

    await run(event);

    expect(sns.publish).not.toBeCalled();
    expect(query).toBeCalledTimes(1);
  });
});
