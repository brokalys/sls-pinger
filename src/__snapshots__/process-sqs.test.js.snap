// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process-sqs works with multiple PINGERs w/ multi-invocation frequencies 1`] = `
Object {
  "area": 45,
  "calc_price_per_sqm": 122.32,
  "category": "apartment",
  "content": "",
  "images": Array [],
  "is_premium": true,
  "lat": 56.9965,
  "lng": 24.245176,
  "price": "100,000",
  "rooms": 2,
  "type": "sell",
  "unsubscribe_url": "https://unsubscribe.brokalys.com/?key=undefined&id=111111",
  "url": "https://view.brokalys.com/?link=https%3A%2F%2Fbrokalys.com%2F",
}
`;
